src/psyclone/psyGen.py

```diff
@@ -988,8 +882,13 @@ def gen_code(self, parent):
         from psyclone.f2pygen import UseGen, DeclGen, AssignGen, CommentGen, \
             IfThenGen, CallGen
 
+        # During gen_code, use a duplicate of the symbol_table so that changes
```

Comment:
This means my comment is not good enough, I will extend it.

The reason is that in the rare case that we do multiple calls of `invoke.gen_code()` (I think it can never happen with `psyclone -s script.py file.f90` but it happens in a small number of unit-tests), without the storing/restoring symbtab, each call will generate new symbols. So:
invoke.gen_code() will use a, b, c, ...
invoke.gen_code() will use a_0, b_0, c_0, ...
invoke.gen_code() will use a_1, b_1, c_1, ...

and even worst, some will try to redefine a tag and fail because tags must be unique.
To be honest, each `invoke.gen_code()` would still generate valid code because it is consistent internally. But this is not the behavior some unit-tests expect.

Reply:
Yes, that makes sense, thanks.

Reply:
The explanation text is good as well. However, it raises a question for me. If we are removing the local symbols from the symbol table after the routine has completed, what is to stop another routine from using the same names and for those names to clash?

Reply:
I am not sure I understand the problem.
If the second routine is executed under the same `InvokeSchedule.gen_code()` gen_codes chain it won't clash because it is the same symbol table.
If it is under a second `InvokeSchedule.gen_code()` it will start the whole generation chain from the beginning, and regenerate all symbols.

If some remains are left after a gen_code (e.g. a Reference to a new Symbol, this is a bug anyway because the code would be repeated 2 times in the second  `InvokeSchedule.gen_code()`)

I think the confusing part is that NameSpaces (and SymbolTables now) are used for 2 IR at the same time: names for PSyIR and names for f2pygen. 
Eventually, instead of a gen_code() we will have a `lower to language level PSyIR' and I hope we will handle this better and we won't need this copy/restore.

Reply:
Updated code explains. Thanks.